### PR TITLE
feature/products-filters

### PR DESCRIPTION
**Pull request by AlexandraVoloshyna**
## Describe your changes

- add endpoint for fetching all products

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](https://olga-team.atlassian.net/browse/TP-193)

### **General**

- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [ ] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [ ] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/)

### Backend

- [x] Swagger documentation updated
- [x] Database requests are optimized and not redundant
- [x] Unit tests written
- [x] use ConfigService instead of process.env
- [ ] use transactions if there is a call chain that mutates data in different tables
- [ ] use @index decorator for frequently requested data

### CODE REVIEW

File: src/modules/product/services/product.service.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    queryBuilder
+      .orderBy('product.name', order)
+      .skip(limit ? (page - 1) * limit : 0)
+      .take(limit);
+
+    const [products, totalCount] = await queryBuilder.getManyAndCount();
+
+    return { totalCount, products };
+  }
```
Comment (eugeneiv1) [inline src/modules/product/services/product.service.ts:15]: queryBuilder in service?

File: src/modules/product/dto/filter-products.dto.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import {
+  IsNumber,
+  IsOptional,
+  IsString,
+  Min,
+  ValidateNested,
+} from 'class-validator';
+
+class PriceRangeDto {
+  @IsNumber()
```
Comment (eugeneiv1) [inline src/modules/product/dto/filter-products.dto.ts]: can it be moved outside in type?

File: src/modules/product/product.controller.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  HttpStatus,
+  Query,
+  //  UseGuards,
+} from '@nestjs/common';
+import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
+
+import { ERouteName } from 'src/common/enums/route-name.enum';
+import { ProductEntity } from 'src/entities/product.entity';
+import { ProductsFilterDto } from 'src/modules/product/dto/filter-products.dto';
+// import { JwtAuthGuard } from 'src/modules/auth/guards/jwtAuthGuard';
```
Comment (eugeneiv1) [inline src/modules/product/product.controller.ts]: why commented? mb add TODO if it is for testing purpose?

File: src/modules/user/services/user-seed.service.ts [hunk 4]
```diff
@@ -5,17 +5,17 @@ import { Not } from 'typeorm';
...
-import { products } from 'src/utils/constants/fakeProducts';
+//import { products } from 'src/utils/constants/fakeProducts';
 import { users } from 'src/utils/constants/fakeUsers';
 
 @Injectable()
 export class UserSeederService implements OnModuleInit {
   constructor(
     private readonly userRepository: UserRepository,
-    private readonly productRepository: ProductRepository,
+    // private readonly productRepository: ProductRepository,
```
Comment (eugeneiv1) [inline src/modules/user/services/user-seed.service.ts]: why commented?

File: src/modules/product/product.controller.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+@Controller(ERouteName.PRODUCTS_ROUTE)
+export class ProductController {
+  constructor(private readonly productService: ProductService) {}
+
+  @ApiBearerAuth()
+  @Get(ERouteName.GET_PRODUCTS_ROUTE)
+  @HttpCode(HttpStatus.OK)
+  async getAllProducts(
+    @Query() query: ProductsFilterDto,
+  ): Promise<{ totalCount: number; products: ProductEntity[] }> {
```
Comment (eugeneiv1) [inline src/modules/product/product.controller.ts]: can return type be moved to type?

File: src/modules/product/services/product-seed.service.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+export class ProductSeederService implements OnModuleInit {
+  constructor(private readonly productRepository: ProductRepository) {}
+
+  async onModuleInit() {
+    const isProductsExist = await this.productRepository.find();
+    if (!isProductsExist.length) {
+      await this.productRepository.save(products);
+    }
+  }
+}
```
Comment (eugeneiv1) [inline src/modules/product/services/product-seed.service.ts]: mb all seeds move to some db-seed.service?

File: src/modules/user/services/user.service.ts [hunk 7]
```diff
@@ -160,24 +160,36 @@ export class UserService {
...
+
+    queryBuilder
+      .orderBy('user.full_name', order)
+      .skip(limit ? (page - 1) * limit : 0)
+      .take(limit);
+
+    const [users, totalCount] = await queryBuilder.getManyAndCount();
+
     return { totalCount, users };
   }
```
Comment (eugeneiv1) [inline src/modules/user/services/user.service.ts:162]: same qb in service?

File: src/modules/product/services/product.service.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      color,
+      priceRange,
+      order = 'ASC',
+    } = dto;
+
+    const queryBuilder = this.productRepository.createQueryBuilder('product');
+
+    if (search) {
+      queryBuilder.andWhere(
+        new Brackets((qb) => {
```
Comment (MishaFomenko) [inline src/modules/product/services/product.service.ts]: maybe we should move this to a repository

File: src/modules/product/services/product.service.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+@Injectable()
+export class ProductService {
+  constructor(private readonly productRepository: ProductRepository) {}
+
+  public async getAllProducts(
+    dto: ProductsFilterDto,
+  ): Promise<{ totalCount: number; products: ProductEntity[] }> {
+    const {
+      page = 1,
+      limit = 5,
```
Comment (MishaFomenko) [inline src/modules/product/services/product.service.ts]: these values could be defined in dto default values

File: src/modules/user/services/user.service.ts (hunk 7 above)
Comment (AlexandraVoloshyna) [reply src/modules/user/services/user.service.ts:162]: moved to user repository

File: src/modules/product/services/product.service.ts (hunk 8 above)
Comment (AlexandraVoloshyna) [reply src/modules/product/services/product.service.ts]: done

**Pull request merged**